Add back keycloak following changes to latest keycloak release

remz1337 opened the pull request:
🛑 **New scripts must first be submitted to [ProxmoxVED](https://github.com/community-scripts/ProxmoxVED) for testing.**  
PRs for new scripts that skip this process will be closed.

---

## ✍️ Description  
Adds keycloak script, with minimal configuration needed with latest release. Update path has also been fixed to avoid issues with `cache-ispn.xml` (see issue: https://github.com/keycloak/keycloak/issues/40046)

## 🔗 Related PR / Issue  

Link: #

## ✅ Prerequisites  (**X** in brackets) 

- [X] **Self-review completed** – Code follows project standards.  
- [X] **Tested thoroughly** – Changes work as expected.  
- [X] **No breaking changes** – Existing functionality remains intact.  
- [X] **No security risks** – No hardcoded secrets, unnecessary privilege escalations, or permission issues.  

---

## 🛠️ Type of Change (**X** in brackets)  

- [ ] 🐞 **Bug fix** – Resolves an issue without breaking functionality.  
- [ ] ✨ **New feature** – Adds new, non-breaking functionality.  
- [ ] 💥 **Breaking change** – Alters existing functionality in a way that may require updates.  
- [X] 🆕 **New script** – A fully functional and tested script or script set.  
- [ ] 🌍 **Website update** – Changes to website-related JSON files or metadata.  
- [ ] 🔧 **Refactoring / Code Cleanup** – Improves readability or maintainability without changing functionality.  
- [ ] 📝 **Documentation update** – Changes to `README`, `AppName.md`, `CONTRIBUTING.md`, or other docs.  

---

## 🔍 Code & Security Review  (**X** in brackets) 

- [X] **Follows `Code_Audit.md` & `CONTRIBUTING.md` guidelines**  
- [X] **Uses correct script structure (`AppName.sh`, `AppName-install.sh`, `AppName.json`)**  
- [ ] **No hardcoded credentials** --> Only for temporary admin bootstrap, as recommended per the official keycloak documentation  


## 📋 Additional Information (optional)  
<!-- Add any extra context, screenshots, or references. -->

MickLesk commented on ct/keycloak.sh:
```diff
@@ -0,0 +1,72 @@
+    apt-get -y upgrade &>/dev/null
+	msg_ok "Updated packages"
+
+    RELEASE=$(curl -fsSL https://api.github.com/repos/keycloak/keycloak/releases/latest | grep "tag_name" | awk '{print substr($2, 2, length($2)-3) }')
+    msg_info "Updating ${APP} to v$RELEASE"
+    cd /opt
+    wget -q https://github.com/keycloak/keycloak/releases/download/$RELEASE/keycloak-$RELEASE.tar.gz
+    mv keycloak keycloak.old
+    tar -xzf keycloak-$RELEASE.tar.gz
+    tar -czf keycloak_conf_backup.tar.gz keycloak.old/conf
+    mv keycloak_conf_backup.tar.gz keycloak-$RELEASE/conf
+    cp -r keycloak.old/providers keycloak-$RELEASE
+    cp -r keycloak.old/themes keycloak-$RELEASE
+    mv keycloak-$RELEASE keycloak
+    rm keycloak-$RELEASE.tar.gz
+    rm -rf keycloak.old
+	msg_ok "Updated ${APP} LXC"
```
the whole path can be simplified with following

```bash
msg_info "Backup old Keycloak"
cd /opt
mv keycloak keycloak.old
tar -czf keycloak_conf_backup.tar.gz keycloak.old/conf
msg_ok "Backup done"

fetch_and_deploy_gh_release "keycloak" "keycloak/keycloak" "tarball" "latest" "/opt/keycloak"

msg_info "Updating ${APP}"
mv keycloak_conf_backup.tar.gz keycloak-$RELEASE/conf
cp -r keycloak.old/providers keycloak-$RELEASE
cp -r keycloak.old/themes keycloak-$RELEASE
rm -rf keycloak.old
msg_ok "Updated ${APP} LXC"

MickLesk commented on install/keycloak-install.sh:
```diff
@@ -0,0 +1,90 @@
+update_os
+
+msg_info "Installing Dependencies (Patience)"
+$STD apt-get install -y curl
+$STD apt-get install -y sudo
+$STD apt-get install -y mc
```
this 3 can be removed (core deps)

MickLesk commented on install/keycloak-install.sh:
```diff
@@ -0,0 +1,90 @@
+$STD apt-get install -y ca-certificates-java
+msg_ok "Installed Dependencies"
+
+msg_info "Installing OpenJDK"
+$STD apt install wget lsb-release -y
+$STD wget https://packages.microsoft.com/config/debian/$(lsb_release -rs)/packages-microsoft-prod.deb -O packages-microsoft-prod.deb
+$STD dpkg -i packages-microsoft-prod.deb
+$STD apt update
+$STD apt install -y msopenjdk-21
+sudo update-java-alternatives --set msopenjdk-21-amd64
+rm packages-microsoft-prod.deb
+msg_ok "Installed OpenJDK"
```
is there anything against adoptium? We actually use it consistently with all refactors and that would only be one line in the installer: 

```bash
JAVA_VERSION=21 setup_java
```

MickLesk commented on install/keycloak-install.sh:
```diff
@@ -0,0 +1,90 @@
+msg_ok "Installed OpenJDK"
+
+msg_info "Installing PostgreSQL"
+$STD apt-get install -y postgresql
```
rm this line, add above (after install dependencies)
```bash
PG_VERSION=16 setup_postgresql
```

MickLesk commented on install/keycloak-install.sh:
```diff
@@ -0,0 +1,90 @@
+$STD sudo -u postgres psql -c "CREATE DATABASE $DB_NAME WITH OWNER $DB_USER ENCODING 'UTF8';"
+$STD sudo -u postgres psql -c "GRANT ALL PRIVILEGES ON DATABASE $DB_NAME TO $DB_USER;"
+msg_ok "Installed PostgreSQL"
+
+msg_info "Installing Keycloak"
+temp_file=$(mktemp)
+RELEASE=$(curl -fsSL https://api.github.com/repos/keycloak/keycloak/releases/latest | grep "tag_name" | awk '{print substr($2, 2, length($2)-3) }')
+curl -fsSL "https://github.com/keycloak/keycloak/releases/download/$RELEASE/keycloak-$RELEASE.tar.gz" -o "$temp_file"
+tar xzf $temp_file
+mv keycloak-$RELEASE /opt/keycloak
+msg_ok "Installed Keycloak"
```
add before msg_info following:

```bash
fetch_and_deploy_gh_release "keycloak" "keycloak/keycloak" "tarball" "latest" "/opt/keycloak"
```

so whole msg_info part is not needed and can be removed

MickLesk commented on install/keycloak-install.sh:
```diff
@@ -0,0 +1,90 @@
+customize
+
+msg_info "Cleaning up"
+rm -f $temp_file
```
can be removed

remz1337 commented:
Thanks for the review @MickLesk , I have addressed the comments.

- I had to change some URLs to point to my fork to test properly. Should I point those back to this repo? Or is it going to be fixed when merging into the main repo (ProxmoxVE)?
- I've used the following pattern `keycloak-*.tar.gz` to get the keycloak binaries using the `fetch_and_deploy_gh_release`. It's working, but there's a risk that it catches a different file from the release (like `keycloak-admin-26.3.0.tar.gz`). Might need to find another solution for that

MickLesk commented:
ive done some little improvements. with this you can do an PR against VE (the file exist)

ive removed this lines:
#echo -e "${TAB}${GN}Temporary admin user:${BL}tmpadm${CL}"
#echo -e "${TAB}${GN}Temporary admin password:${BL}admin123${CL}"
#echo -e "${INFO}${YW} If you modified ${BL}cache-ispn.xml${YW}: Re-apply your changes to the new file, otherwise leave it unchanged.${CL}"

=> maybe add this into the json (the user/admin as username / password) and the last as "info" note
